[Feature][Kernel] Support bitsandbytes quantization and QLoRA

chenqianfzh opened the pull request:
QLoRA (https://arxiv.org/abs/2305.14314) cuts memory consumption in LLM weight loading without degrading performance. The weights of the basic model , which are quantized into 4 bit using bitsandbytes quantization,  pair with a low-rank but higher-precision Low-Rank weight matrix to generate output. 

This MR is the first step in supporting QLoRA in vLLM. With the PR,  the Qlora author's open model on hugging face, such as, is supported:
  - https://huggingface.co/timdettmers/qlora-flan-7b (Its corresponding large model is "huggyllama/llama-7b")

User can run with or without a QLoRA adapter.

So far, only llama as a basic model is supported. More to come in the future. As explained below, special consideration is made for extensibility to future changes and other models.
Also, no TP or PP with QLoRA is supported. It will be considered as the immediate next effort.
  

**Explanation on Changes**

Modified files mainly include
- Modify vllm/config.py, vllm/engine/arg_utils.py: Add new CLI parameters for QLoRA/bitsandbytes. The new parameter is:
	- qlora_adapter_name_or_path : the path to the adpater repo. Could be empty.
- Modify vllm/model_loader/loader.py: Define a new loader class, which will quantize the weight using bitsandbytes during loading
- Modify vllm/model_executor/layers/linear.py: Add the logic of concatenate tensor in bitsandbytes in the weight_loader () function of QKVParallelLinear class and MergedColumnParallelLinear class.

The newly added files are:
- VLLM/model_executor/layers/quantization/bitsandbytes.py: Here, similar to other quantization methods, we define two classes, class BitsAndBytesConfig (QuantizationConfig) and class BitsAndBytesLinearMethod (LinearMethodBase). 
- Examples/qlora_inference.py: Demonstration of the use of bitsandbytes, both with and without an adapter.

jeejeelee commented:
ping @Yard1

Yard1 reviewed (changes requested):
This completely bypasses the existing LoRA logic and implements its own. I don't think this is a good design and it clashes with already existing code. We should instead modify the LoRA support already present in vLLM to support QLoRA - it should also allow us to reuse a lot of existing code.

chenqianfzh commented:
> This completely bypasses the existing LoRA logic and implements its own. I don't think this is a good design and it clashes with already existing code. We should instead modify the LoRA support already present in vLLM to support QLoRA - it should also allow us to reuse a lot of existing code.

Thanks for your reply. You are not the first one who popped this concern. Actually, I asked myself the same question. :-)

I considered about re-use of LoRA at the first place. I have to start a new set of code because:
1. the existing LoRA in vLLM is implementing punica (https://github.com/punica-ai/punica), a multi-tenant scenario of LoRA. A lot of effort have made on the LoRA manager, which manages the cases where different sets of fine-tune weights using the same basic models. 

But QLoRA, though carring a very similar name, work for a totally different scenario, thus unable to re-use the existing code of LoRA in vLLM.

2. punica is based on cuda code of BGMV, and BGMV does not support any quantization. But in QLORA, quantization of basic model is the keypoint in saving memory. This is another reason I had to deviate away from reusing LoRA. 

3. On the other hand, QLoRA use a different set of cuda code. The author of QLORA provides the Cuda implementation of QLORA implemention and packed in the python package of bitsandbytes, which is used in the QLORA implementation of huggingface transformers package.  So I moved away from re-using the LoRA code.

How about I add some comments somewhere to clarify your concern?

Yard1 commented:
Is it possible for the adapter to be loaded and unloaded at will?

chenqianfzh commented:
> Is it theoretically possible for the QLoRA adapter to be loaded and unloaded at will?

I am not sure what you mean by "at will". Do you mean load/unload during runtime?

In this implementation, user can load an adpater by specfiying "qlora_adapter_name_or_path" in parameter when starting the inference. User can also run without an adapter by leaving the above parameter empty.

However, the user cannot switch the adapter during the runtime. Switching adapter is not a scenario supported in the QLoRA design.

The main goal of QLoRA is to use to LoRA weights to compensate the loss caused by the 4-bit quantization in the basic model. So it is a quantization technique. Switching LoRA to support different fine-tune scenarios, as in punica, is not in its design goals.

Yard1 commented:
Ok, that's what I wanted to confirm. Thanks for clearing it up. In that case:
1. for consistency, I would suggest ditching the `qlora_supported` decorator and just specify the class attribute directly on the class
2. we should avoid the `if model_config.quantization == "qlora":` pattern in linear layer and weight loading code - instead we should use abstractions (and add them if they are missing). For example, we should add a `QLoRAModelLoader` which can subclass/compose `DefaultModelLoader`. Same for linear layer - we should avoid adding special cases to generic implementations (I understand this pattern is not always followed in the codebase, but we should hold new code to higher standard - happy to discuss what sort of API we need to add to get rid of the `Special case for Quantized Weights.` in linear layer implementation)

chenqianfzh commented:
> Ok, that's what I wanted to confirm. Thanks for clearing it up. In that case:
> 
> 1. for consistency, I would suggest ditching the `qlora_supported` decorator and just specify the class attribute directly on the class
> 2. we should avoid the `if model_config.quantization == "qlora":` pattern in linear layer and weight loading code - instead we should use abstractions (and add them if they are missing). For example, we should add a `QLoRAModelLoader` which can subclass/compose `DefaultModelLoader`. Same for linear layer - we should avoid adding special cases to generic implementations (I understand this pattern is not always followed in the codebase, but we should hold new code to higher standard - happy to discuss what sort of API we need to add to get rid of the `Special case for Quantized Weights.` in linear layer implementation)

Thanks for the suggestion. I will make the changes as suggested.

Cheers!

jeejeelee commented:
Thank you for your excellent work. Here are some personal opinions:
- vLLM has supported quantized models with LoRA, refer to [quant model+lora](https://github.com/vllm-project/vllm/blob/main/tests/lora/test_quant_model.py). These can be generalized as QLoRA (e.g., GPTQ+LoRA), and all of them support switching adapters.
- For the original QLoRA (https://arxiv.org/abs/2305.14314), I think we should add a new quantization method named `bitsandbytes` (e.g., BAB+LoRA), refer to https://github.com/vllm-project/vllm/issues/4033, and then we can reuse the current LoRA logic.
- Regardless of LoRA or QLoRA, Punica can support these

If I am wrong, please correct me directly, Thanks again.

Cheers!

chenqianfzh commented:
> Thank you for your excellent work. Here are some personal opinions:
> 
> * vLLM has supported quantized models with LoRA, refer to [quant model+lora](https://github.com/vllm-project/vllm/blob/main/tests/lora/test_quant_model.py). These can be generalized as QLoRA (e.g., GPTQ+LoRA), and all of them support switching adapters.
> * For the original QLoRA (https://arxiv.org/abs/2305.14314), I think we should add a new quantization method named `bitsandbytes` (e.g., BAB+LoRA), refer to [[Feature]: bitsandbytes support #4033](https://github.com/vllm-project/vllm/issues/4033), and then we can reuse the current LoRA logic.
> * Regardless of LoRA or QLoRA, Punica can support these
> 
> If I am wrong, please correct me directly, Thanks again.
> 
> Cheers!

I re-read the LoRA code carefully and saw that quantization is supported in LoRA now, which change was checked in 4/27. 

I will re-think my design again based on this change, as well as Yard1's suggestions. 

Thanks & Happy Coding!

chenqianfzh commented:
@Yard1 @jeejeelee 

I just updated the MR of QLoRA/BitsAndBytes with the changes suggested. Could you please take another look? 

Thanks for the great advice from you. Learned a lot and improved a lot. :-)

BTW, I hit a lot of yapf errors in CI/CD. I found the the yapf errors are not from me. Should I just ignore it?

jeejeelee commented:
@chenqianfzh  We cannot igore format error, you can run `bash format.sh` to check for format errors

Yard1 commented on vllm/model_executor/model_loader/loader.py:
```diff
@@ -526,6 +533,201 @@ def save_model(
 
 
+class BitsAndBytesModelLoader(BaseModelLoader):
+    """Model loader that will set model weights to random values."""
```
fix docstring

chenqianfzh replied:
Fixed.

Yard1 commented on vllm/model_executor/model_loader/loader.py:
```diff
@@ -526,6 +533,201 @@ def save_model(
+
+    def __init__(self, load_config: LoadConfig):
+        super().__init__(load_config)
+        default_target_modules = [
+            "gate_proj", "down_proj", "up_proj", "q_proj", "k_proj", "v_proj",
+            "o_proj"
+        ]
```
let's make it a class attribute. also isn't it better to always just read from the config? is this a common way to handle unspecified layers? not familiar with this myself!

chenqianfzh replied:
BnB can work without a LoRA adpater. Such an example was given in hugging face blog, https://huggingface.co/blog/4bit-transformers-bitsandbytes. 

What i am doing here is following the handling of BitsAndBytesConfig for Llama model in hugging face transformer package. Namely, only the listed modules will be quantized, while other modules, such as lm_head, embed_tokens, layer_norms, are not quantized. 

Since no quant_config files to read if there is not LoRA adapter specified, I have to put them as hard-coded. 

However, I have to say, the list of default target modules may be different for other models. But since only Llama is supported in this MR, I just put it as is. I will change if necessary, that is for sure.

Yard1 commented on vllm/model_executor/model_loader/loader.py:
```diff
@@ -526,6 +533,201 @@ def save_model(
+        is_local = os.path.isdir(qlora_adapter)
+        config_file_name = ["adapter_config.json"]
+        config_file_path = None
+        if is_local:
+            for file in config_file_name:
+                config_file_path = os.path.join(qlora_adapter, file)
+                if os.path.exists(config_file_path):
+                    break
+        else:
+            hf_api = HfApi()
+            repo_files = hf_api.list_repo_files(repo_id=qlora_adapter)
+            for file in config_file_name:
+                if file in repo_files:
+                    config_file_path = hf_hub_download(repo_id=qlora_adapter,
+                                                       filename=file)
+                    break
```
let's separate this out into a generic function for better code reuse (`maybe_download_from_hf`)

chenqianfzh replied:
updated as suggested.

Yard1 commented on vllm/model_executor/model_loader/loader.py:
```diff
@@ -526,6 +533,201 @@ def save_model(
+        allowed_patterns = ["*.safetensors", "*.bin", "*.pt"]
+        is_local = os.path.isdir(model_name_or_path)
+
+        if not is_local:
+            hf_api = HfApi()
+            repo_files = hf_api.list_repo_files(repo_id=model_name_or_path)
+            for pattern in allowed_patterns:
+                matching_files = fnmatch.filter(repo_files, pattern)
+                if len(matching_files) > 0:
+                    matched_pattern = pattern
+                    break
+
+            if not matched_pattern:
+                raise RuntimeError(
+                    f"Cannot find any model weights with `{model_name_or_path}`"
+                )
+
+            hf_folder = download_weights_from_hf(model_name_or_path,
+                                                 self.load_config.download_dir,
+                                                 [matched_pattern], revision)
+            hf_weights_files = glob.glob(
+                os.path.join(hf_folder, matched_pattern))
+        else:
+            hf_folder = model_name_or_path
+            for pattern in allowed_patterns:
+                hf_weights_files = glob.glob(os.path.join(hf_folder, pattern))
+                if len(hf_weights_files) > 0:
+                    matched_pattern = pattern
+                    break
```
separate this out as well

chenqianfzh replied:
updated.

Yard1 commented on vllm/engine/arg_utils.py:
```diff
@@ -550,6 +558,27 @@ def from_cli_args(cls, args: argparse.Namespace):
+                        f"BitsAndBytes quantization cannot work with {self.load_format} load format"
+                    )
+
+        if self.load_format == "bitsandbytes" or self.qlora_adapter_name_or_path is not None:
+            if self.quantization == None:
+                self.quantization = "bitsandbytes"
+            else:
+                if self.quantization != "bitsandbytes":
+                    raise ValueError(
+                        f"BitsAndBytes load format cannot work with {self.quantization} "
+                    )
```
let's just throw an exception if load format is bitsandbytes but quantization is not (and tell the user to fix it themselves)

chenqianfzh replied:
fixed.

Yard1 commented on vllm/config.py:
```diff
@@ -443,6 +449,7 @@ class LoadFormat(str, enum.Enum):
     DUMMY = "dummy"
     TENSORIZER = "tensorizer"
     SHARDED_STATE = "sharded_state"
+    BitsAndBytes = "bitsandbytes"
```
use all capital letters for enum values

Yard1 commented on vllm/model_executor/layers/quantization/bitsandbytes.py:
```diff
@@ -0,0 +1,175 @@
+        if not isinstance(pack_ratio, int):
+            raise RuntimeError("Pack ratio is not an integer.")
+
+        num_elements = [0] * len(quant_states)
+        for seq, quant_state in enumerate(quant_states.items()):
+            num_elements[seq] = math.prod(quant_state[1].shape) // pack_ratio
+
+        offsets = np.concatenate(([0], np.cumsum(num_elements)))
```
this will be inefficient. We should do the following to optimize:
1. move all the checks to `create_weights`/`__init__`, we cannot have them on the hot path
2. as far as I can see, `offsets` will be a per-layer constant. pre-calculate it in `create_weights`, set it on `extra_weight_attrs` and reuse multiple times

chenqianfzh replied:
fixed.

Yard1 commented on vllm/model_executor/layers/quantization/bitsandbytes.py:
```diff
@@ -0,0 +1,175 @@
+
+        offsets = np.concatenate(([0], np.cumsum(num_elements)))
+
+        out = matmul_4bit(bf_x, qweight[offsets[0]:offsets[1]].t(),
+                          quant_states[0])
+
+        for i in range(1, len(quant_states)):
+
+            current_basic_output = matmul_4bit(
+                bf_x, qweight[offsets[i]:offsets[i + 1]].t(), quant_states[i])
+
+            out = torch.cat((out, current_basic_output), dim=-1)
```
Since we know the final shape of `out` ahead of time, we should do the following for best performance:
1. preallocate the entire `out` tensor with `torch.empty`
2. use the `out` kwarg of `matmul_4bit` to directly write data to slices of preallocated tensor

that way we can avoid having to concatenate in the end, which is inefficient (tip: if you cannot preallocate the tensor, then you should instead save the shards into a list and do one `torch.cat` call at the end)

also can we put the first case into the loop as well?

chenqianfzh replied:
refactored as suggested. Thanks for the insightful advice.

chenqianfzh replied:
updated. Thanks for the suggestion

Yard1 commented on vllm/model_executor/layers/quantization/bitsandbytes.py:
```diff
@@ -0,0 +1,175 @@
+        out = out.to(orginal_type)
+
+        if bias is not None:
+            out = out + bias
```
```suggestion
            out += bias
```
this way we do an in-place addition which saves memory

chenqianfzh replied:
Thanks for catching this.

chenqianfzh replied:
updated.

Yard1 commented on vllm/model_executor/model_loader/loader.py:
```diff
@@ -526,6 +533,201 @@ def save_model(
+                       for target_module in self.target_modules):
+                    weight_name = weight_name.replace(".weight", ".qweight")
+                    loaded_weight = weight_tensor.cuda().data  # Move to GPU
+                    torch.set_default_dtype(torch.float32)
```
Use the `set_default_torch_dtype` context manager for setting the default dtype (otherwise, if there is an exception, we are left with the wrong default dtype being set)

chenqianfzh replied:
updated

Yard1 commented on vllm/model_executor/model_loader/loader.py:
```diff
@@ -526,6 +533,201 @@ def save_model(
+                if any(target_module in weight_name
+                       for target_module in self.target_modules):
+                    weight_name = weight_name.replace(".weight", ".qweight")
+                    loaded_weight = weight_tensor.cuda().data  # Move to GPU
```
move to device instead? if bnb only supports cuda, we should throw an exception much earlier if the device doesn't match (remember vllm supports multiple accelerator types)

chenqianfzh replied:
bnb quantization requires the data to be quantized to be in GPU. The weight_tensor are the weights just read from the weight files downloaded from hugging face, so it can only be in CPU. Therefore, I explicitly have to move them GPU explicitly.

Yard1 reviewed (commented):
Thanks, this is looking *much* cleaner! Left some comments, hope they will be useful.

Yard1 commented:
We should also add a test for this - it's ok if it's just an end to end one (load a small model from huggingface hub and see if it works and gives good outputs)

mgoin commented on requirements-common.txt:
```diff
@@ -19,3 +19,4 @@ lm-format-enforcer == 0.10.1
 outlines == 0.0.34 # Requires torch >= 2.1.0
 typing_extensions
 filelock >= 3.10.4 # filelock starts to support `mode` argument from 3.10.4
+bitsandbytes == 0.42.0
```
This doesn't belong in the common requirements and actually I think we shouldn't require bitsandbytes by default in the package. I think in `vllm/model_executor/layers/quantization/bitsandbytes.py` we should check if it is installed and throw then. This is what was done for `deepspeedfp` https://github.com/vllm-project/vllm/blob/606625329648e6eff1883e23040adfad82f219cf/vllm/model_executor/layers/quantization/deepspeedfp.py#L145-L154

chenqianfzh replied:
fixed.

chenqianfzh commented:
@mgoin @Yard1 @jeejeelee 

Thanks for the feedback. Working on the changes now.

chenqianfzh commented:
> We should also add a test for this - it's ok if it's just an end to end one (load a small model from huggingface hub and see if it works and gives good outputs)

the newly added file examples/qlora_inference.py is created for this purpose. In this file, both the case that bitsandbytes quantization with/withou LoRA adpaters are tested.

Here are the ouput I got in my local test ( of the four, the last is without a LORA adapter , the other three are with adpaters:
```
--------------------------------------------------------------------------
Prompt: The capital of France is 
Output:  Paris.
--------------------------------------------------------------------------
Prompt: The capital of USA is 
Output:  Washington DC.
--------------------------------------------------------------------------
Prompt: my name is 
Output:  john and i am a 20 year old male. i am a student at the university of maryland. i am a sophomore and i am majoring in business. i am a very outgoing person and i love to meet new people. i am a very social person and i love to party. i am a very outgoing person and i love to meet new people. i am a very social person and i love to party.
--------------------------------------------------------------------------
Prompt: My name is 
Output:  Kyle and I am a 20 year old college student. I am a huge fan of the outdoors and love to hike, camp, and fish. I am a very active person and love to stay busy. I am a very outgoing person and love to meet new people. I am a very easy going person and love to have fun. I am a very hard worker and love to work. I am a very trustworthy person and love to help people. I am a very caring person and love to help people. I am a very respectful person and love to respect others. I am a

```

jeejeelee commented:
@chenqianfzh  Can we add more quantization type examples in qlora_example.py, such as GPT+LoRA, so that users can refer to this script to learn how to utilize LoRA on quantized model, thanks

chenqianfzh commented:
@jeejeelee @Yard1 @mgoin 

I have updated the PR, addressing and resolving all the comments. Additionally, I have added the necessary unit tests. Please review it again.

However, I was unable to "add more quantization type examples in qlora_example.py" at this time. Currently, Llama is the only model supported in this MR. Expanding support to more models is my next task. As suggested by @jeejeelee, GPT will likely be the next model to support, given the availability of several GPT-based Qlora models, such as https://huggingface.co/vineetsharma/qlora-gpt-neox-20b-english_quotes/

It's important to note that adding support for GPT may require additional effort because VLLM does not currently support LoRA in GPT-Neox. Therefore, I believe it would be more appropriate to address this in a separate PR.

Thank you again for all the great suggestions.

jeejeelee commented:
> @jeejeelee @Yard1 @mgoin
> 
> I have updated the PR, addressing and resolving all the comments. Additionally, I have added the necessary unit tests. Could u please review it again?
> 
> However, I was unable to "add more quantization type examples in qlora_example.py" at this time. Currently, Llama is the only model supported in this MR. Expanding support to more models is my next task. As suggested by @jeejeelee, GPT will likely be the next model to support, given the availability of several GPT-based Qlora models, such as https://huggingface.co/vineetsharma/qlora-gpt-neox-20b-english_quotes/
> 
> It's important to note that adding support for GPT may require additional effort because VLLM does not currently support LoRA in GPT-Neox. Therefore, I believe it would be more appropriate to address this in a separate PR.
> 
> Thank you again for all the great suggestions.

My apologies, there was an issue with my spelling. What I actually meant to say was GPTQ+LORA. I'm very sorry.

chenqianfzh commented:
> > @jeejeelee @Yard1 @mgoin
> > I have updated the PR, addressing and resolving all the comments. Additionally, I have added the necessary unit tests. Could u please review it again?
> > However, I was unable to "add more quantization type examples in qlora_example.py" at this time. Currently, Llama is the only model supported in this MR. Expanding support to more models is my next task. As suggested by @jeejeelee, GPT will likely be the next model to support, given the availability of several GPT-based Qlora models, such as https://huggingface.co/vineetsharma/qlora-gpt-neox-20b-english_quotes/
> > It's important to note that adding support for GPT may require additional effort because VLLM does not currently support LoRA in GPT-Neox. Therefore, I believe it would be more appropriate to address this in a separate PR.
> > Thank you again for all the great suggestions.
> 
> My apologies, there was an issue with my spelling. What I actually meant to say was GPTQ+LORA. I'm very sorry.

@jeejeelee I see. Yeah, I can do that.

I see you are the one who made LoRA support quant methods. Could you let me know your suggestions on what models to show the usage of GPTQ+LORA?

jeejeelee commented:
> > > @jeejeelee @Yard1 @mgoin
> > > I have updated the PR, addressing and resolving all the comments. Additionally, I have added the necessary unit tests. Could u please review it again?
> > > However, I was unable to "add more quantization type examples in qlora_example.py" at this time. Currently, Llama is the only model supported in this MR. Expanding support to more models is my next task. As suggested by @jeejeelee, GPT will likely be the next model to support, given the availability of several GPT-based Qlora models, such as https://huggingface.co/vineetsharma/qlora-gpt-neox-20b-english_quotes/
> > > It's important to note that adding support for GPT may require additional effort because VLLM does not currently support LoRA in GPT-Neox. Therefore, I believe it would be more appropriate to address this in a separate PR.
> > > Thank you again for all the great suggestions.
> > 
> > 
> > My apologies, there was an issue with my spelling. What I actually meant to say was GPTQ+LORA. I'm very sorry.
> 
> @jeejeelee I see. Yeah, I can do that.
> 
> I see you are the one who made LoRA support quant methods. Could you let me know your suggestions on what models to show the usage of GPTQ+LORA?

Thanks ,you can refer to [test_quant_model.py](https://github.com/vllm-project/vllm/blob/main/tests/lora/test_quant_model.py)

Yard1 commented:
Thanks! In the future, please avoid force pushing as it makes reviews harder.

chenqianfzh commented:
> Thanks! In the future, please avoid force pushing as it makes reviews harder.

I will try to. However, as the main branch is moving really fast, I need to rebase from time to time and I have to force push after rebase. 

Do you mean that I should keep the changes based on comments as a separate commit?

Yard1 commented on tests/quantization/test_bitsandbytes.py:
```diff
@@ -0,0 +1,57 @@
+@pytest.mark.skipif(
+    capability < QUANTIZATION_METHODS["bitsandbytes"].get_min_capability(),
+    reason="bitsandbytes is not supported on this GPU type.")
+def test_load_bnb_model(vllm_runner) -> None:
```
can we also add an e2e test that will compare the output of some model to predefined outputs?

chenqianfzh replied:
sure, added.

Yard1 commented on vllm/config.py:
```diff
@@ -229,6 +229,12 @@ def verify_with_parallel_config(
+                parallel_config.tensor_parallel_size > 1
+                or parallel_config.pipeline_parallel_size > 1):
+            raise ValueError(
+                "BitAndBytes with TP or PP is not supported yet. Will do.")
```
```suggestion
                "BitsAndBytes quantization with TP or PP is not supported yet.")
```

Yard1 commented on vllm/engine/arg_utils.py:
```diff
@@ -555,6 +563,20 @@ def from_cli_args(cls, args: argparse.Namespace):
+        if (self.quantization == "bitsandbytes" or
+            self.qlora_adapter_name_or_path is not None) and \
+            self.load_format != "bitsandbytes":
+            raise ValueError("BitsAndBytes quantization cannot work with "
+                             f"{self.load_format} load format")
```
```suggestion
            raise ValueError("BitsAndBytes quantization cannot work with "
                             f"only supports "bitsandbytes", load format got {self.load_format}")
```

chenqianfzh replied:
updated.

Yard1 commented on vllm/engine/arg_utils.py:
```diff
@@ -555,6 +563,20 @@ def from_cli_args(cls, args: argparse.Namespace):
+        if (self.load_format == "bitsandbytes" or
+            self.qlora_adapter_name_or_path is not None) and \
+            self.quantization != "bitsandbytes":
+            raise ValueError("BitsAndBytes load format cannot work with "
+                             f"{self.quantization}")
```
```suggestion
            raise ValueError("BitsAndBytes load format can only be used with "
                             f"'bitsandbytes' quantization, got {self.quantization}")
```

chenqianfzh replied:
updated.

Yard1 commented on vllm/model_executor/layers/linear.py:
```diff
@@ -26,6 +26,17 @@ def adjust_marlin_shard(param, shard_size, shard_offset):
     return shard_size * marlin_tile_size, shard_offset * marlin_tile_size
 
 
+def adjust_bitsandbytes_shard(param, qkv_offsets, loaded_shard_id):
```
add docstring and type hints

chenqianfzh replied:
added.

chenqianfzh committed:
[model] support bitsandbytes/QLoRA (#4033)

chenqianfzh committed:
Merge branch 'main' into qian/qlora

chenqianfzh commented:
@jeejeelee @Yard1 

The PR was updated per your comments. Thanks for reviewing it. Could you take another look?

There are some CI failures, but it looks like someone is fixing it. 

I tried to merge instead of rebase. However, I see an extra merge commit in my private branch. Any idea what has gone wrong?

Thanks.

Yard1 commented:
@chenqianfzh the merge commit is expected, that's just how git works

chenqianfzh commented:
> @chenqianfzh the merge commit is expected, that's just how git works

I did something wrong in squashing commits before merging, so the commits are mixed today. Sorry to make your review more difficult. :-(

chenqianfzh committed:
Merge branch 'main' into qian/qlora

Yard1 commented on vllm/engine/arg_utils.py:
```diff
@@ -555,6 +563,23 @@ def from_cli_args(cls, args: argparse.Namespace):
+
+        # bitsandbytes quantization needs a specific model loader
+        # so we make sure the quant method and the load format are consistent
+        if (self.quantization == "bitsandbytes" or
+            self.qlora_adapter_name_or_path is not None) and \
+            self.load_format != "bitsandbytes":
+            raise ValueError(
+                "BitsAndBytes quantization only supports 'bitsandbytes' load "
+                f"format, but got {self.load_format}")
+
+        if (self.load_format == "bitsandbytes" or
+            self.qlora_adapter_name_or_path is not None) and \
+            self.quantization != "bitsandbytes":
+            raise ValueError(
+                "BitsAndBytes load format can only be used with bitsandbytes'"
+                f" quantization, but got {self.quantization}")
```
nit: let's throw exception if `self.quantization != "bitsandbytes" and self.qlora_adapter_name_or_path is not None`

chenqianfzh replied:
The case of 'self.quantization != "bitsandbytes" and self.qlora_adapter_name_or_path is not None', actually is already covered in the code from line 576 to 579.

So I rephrased the error messages to mention the case that "qlora_adapter_name_or_path is not None" as well.

Yard1 commented on vllm/model_executor/layers/quantization/bitsandbytes.py:
```diff
@@ -0,0 +1,171 @@
+        current_index = 0
+        for i in range(len(quant_states)):
+            output_size = quant_states[i].shape[0]
+            out[:, current_index:current_index + output_size] = matmul_4bit(
+                bf_x, qweight[offsets[i]:offsets[i + 1]].t(), quant_states[i])
```
```suggestion
             matmul_4bit(
                bf_x, qweight[offsets[i]:offsets[i + 1]].t(), quant_states[i], out=out[:, current_index:current_index + output_size])
```

chenqianfzh replied:
I remember I tried the way you suggested before, it did not work as expected, the out tensor was not set. I tried it again today and confirmed it. 

I think it is a bug in bitsandbytes. :-(

Yard1 replied:
damn, ok. can you open an issue in bugsandbytes and link it in the comment here saying this is why we can't use the `out` kwarg

chenqianfzh replied:
yep, updated the comments as suggested.

Yard1 reviewed (commented):
Thanks, left two last nits! We can merge after those are resolved.

chenqianfzh committed:
update error messages

chenqianfzh committed:
Merge branch 'main' into qian/qlora

chenqianfzh commented:
> Thanks, left two last nits! We can merge after those are resolved.

I've updated the code based on your feedback and have omitted one comment, for which I've provided an explanation. Could you please take a look?

thanks.

chenqianfzh committed:
Merge branch 'main' into qian/qlora

chenqianfzh committed:
Merge branch 'main' into qian/qlora

chenqianfzh commented:
@Yard1 I kept trying the CI tests in the past two days. But hit all kinds of weird errors, like the latest failure is due to a container missing in AMD tests. 

I did not find a way to restart the specific tests. Could you let me know what to do? Thanks.

Yard1 commented:
It's OK, we'll just have a maintainer force merge it

chenqianfzh committed:
add comments about bitandbytes bug

chenqianfzh committed:
Merge branch 'main' into qian/qlora

chenqianfzh committed:
Revert "add comments about bitandbytes bug"

This reverts commit 161c792556001d50b4e37f73603aa5346536850a.

chenqianfzh committed:
add comment about bitsandbytes bug

Yard1 commented on examples/offline_inference.py:
```diff
@@ -11,7 +11,7 @@
 
 # Create an LLM.
-llm = LLM(model="facebook/opt-125m")
+llm = LLM(model="facebook/opt-125m", pipeline_parallel_size=2)
```
Revert

chenqianfzh replied:
done.

chenqianfzh committed:
Merge branch 'main' into qian/qlora

mgoin commented on vllm/model_executor/layers/quantization/bitsandbytes.py:
```diff
@@ -0,0 +1,175 @@
+        from bitsandbytes import matmul_4bit
+
+        orginal_type = x.dtype
+        bf_x = x.to(torch.bfloat16)
```
Based on this, I think get_supported_act_dtypes should include torch.bfloat16

chenqianfzh replied:
Thanks for catching this. I check bitsandbytes and found bfloat16 and float32 are also supported. I fixed the code.

mgoin commented on vllm/config.py:
```diff
@@ -241,6 +241,12 @@ def verify_with_parallel_config(
+        if self.quantization == "bitsandbytes" and (
+                parallel_config.tensor_parallel_size > 1
+                or parallel_config.pipeline_parallel_size > 1):
+            raise ValueError(
+                "BitAndBytes quantization with TP or PP is not supported yet.")
```
Is this easily done, or is there complexity within the quantization format that makes it difficult? It is worth noting for future work if so

chenqianfzh replied:
I understand that Pipeline Parallelism (PP) is currently disabled in vLLM, so I will not be focusing on it for now.

Tensor Parallelism (TP) should not be overly challenging, but it does require careful calibration of tensor sharding and concatenation, especially since the quantized tensor in bitsandbytes has a different shape than its original one. This is my first pull request for this community, and I aimed to keep it focused.

As I mentioned in the pull request description, parallelism will be my immediate next effort. I look forward to your understanding and support.

mgoin replied:
Thanks for the clarity! Yes, TP would be nice as a followup in the near future

mgoin commented on vllm/model_executor/layers/quantization/bitsandbytes.py:
```diff
@@ -0,0 +1,175 @@
+            qweight,
+            {
+                "input_dim": 0,
+                # In bitsandbytes, a tensor of shape [n,m] is qunatized to
```
```suggestion
                # In bitsandbytes, a tensor of shape [n,m] is quantized to
```

chenqianfzh replied:
thanks for catching this. Fixed.

mgoin commented on vllm/model_executor/layers/quantization/bitsandbytes.py:
```diff
@@ -0,0 +1,175 @@
+        # only load the bitsandbytes module when needed
+        from bitsandbytes import matmul_4bit
+
+        orginal_type = x.dtype
```
orginal_type -> original_type

chenqianfzh replied:
fixed.

mgoin commented on vllm/model_executor/model_loader/loader.py:
```diff
@@ -539,6 +546,242 @@ def save_model(
 
 
+class BitsAndBytesModelLoader(BaseModelLoader):
+    """Model loader to load model weights with BitAndBytes quabntization."""
```
```suggestion
    """Model loader to load model weights with BitAndBytes quantization."""
```

chenqianfzh replied:
fixed.

mgoin commented on vllm/model_executor/model_loader/loader.py:
```diff
@@ -539,6 +546,242 @@ def save_model(
+    ) -> Tuple[Generator[Tuple[str, torch.Tensor], None, None], Dict[str,
+                                                                     Any]]:
+        """Get an iterator to the model weights with bitsandbytes quantization,
+
```
```suggestion
```

chenqianfzh replied:
fixed.

mgoin commented on vllm/model_executor/models/llama.py:
```diff
@@ -319,6 +319,14 @@ class LlamaForCausalLM(nn.Module):
         "lm_head": "output_embeddings",
     }
     embedding_padding_modules = ["lm_head"]
+    bitsandbytes_stacked_params_mapping = {
+        # shard_name, weight_name, index
+        "q_proj": ("qkv_proj", 0),
+        "k_proj": ("qkv_proj", 1),
+        "v_proj": ("qkv_proj", 2),
+        "gate_proj": ("gate_up_proj", 0),
+        "up_proj": ("gate_up_proj", 1),
+    }
```
Does this mean only Llama is supported at the moment and we will need to add this mapping for each model? It is worth noting this needs to be present for each model - maybe the bitsandbytes.py file is a good place to put this information

chenqianfzh replied:
yes, only llama is supported right now. Similar to the feature of LoRA, some changes need to make to each model to support QLoRA. Different models might have slightly different list of module names. 

This variable also serves as a mark to indicate whether QLoRA is supported in the model. It is following the same pattern of LoRA, which use the variable of supported_lora_modules for indication.

chenqianfzh committed:
update per comments

chenqianfzh commented:
@mgoin Thanks for reviewing the PR!

I updated the code per your comments. Could u have another check?

sajadn commented:
Hey, 
Thanks for the feature. 

why do you make sure 'lm_head' is not quantized in your tests while peft accepts 'lm_head' among the target_modules? I was trying to run inference for a model fine-tuned with qlora and I get the following error: 

```
File "/opt/conda/envs/llm/lib/python3.9/site-packages/vllm/model_executor/models/llama.py", line 436, in load_weights
    param = params_dict[name]
KeyError: 'lm_head.qweight'
```

mgoin commented:
Hey @sajadn, it may be related to the precedent of GPTQ not having quantized `lm_head`. See this PR for possibly more context https://github.com/vllm-project/vllm/pull/4442 cc @robertgshaw2-neuralmagic

devlup commented:
hi https://github.com/vllm-project/vllm/pull/4776 @chenqianfzh  it seems the bitsandbytes test is not working with LLama 3 , can you retest

VirgilG72 commented:
Hello, I am appreciated for your excellent work. However, I have noticed about that "Also, no TP or PP with QLoRA is supported. It will be considered as the immediate next effort.". So, May I ask when this job is expected to be supported in the future？

> QLoRA (https://arxiv.org/abs/2305.14314) cuts memory consumption in LLM weight loading without degrading performance. The weights of the basic model , which are quantized into 4 bit using bitsandbytes quantization, pair with a low-rank but higher-precision Low-Rank weight matrix to generate output.
> 
> This MR is the first step in supporting QLoRA in vLLM. With the PR, the Qlora author's open model on hugging face, such as, is supported:
> 
> * https://huggingface.co/timdettmers/qlora-flan-7b (Its corresponding large model is "huggyllama/llama-7b")
> 
> User can run with or without a QLoRA adapter.
> 
> So far, only llama as a basic model is supported. More to come in the future. As explained below, special consideration is made for extensibility to future changes and other models. Also, no TP or PP with QLoRA is supported. It will be considered as the immediate next effort.
> 
> **Explanation on Changes**
> 
> Modified files mainly include
> 
> * Modify vllm/config.py, vllm/engine/arg_utils.py: Add new CLI parameters for QLoRA/bitsandbytes. The new parameter is:
>   
>   * qlora_adapter_name_or_path : the path to the adpater repo. Could be empty.
> * Modify vllm/model_loader/loader.py: Define a new loader class, which will quantize the weight using bitsandbytes during loading
> * Modify vllm/model_executor/layers/linear.py: Add the logic of concatenate tensor in bitsandbytes in the weight_loader () function of QKVParallelLinear class and MergedColumnParallelLinear class.
> 
> The newly added files are:
> 
> * VLLM/model_executor/layers/quantization/bitsandbytes.py: Here, similar to other quantization methods, we define two classes, class BitsAndBytesConfig (QuantizationConfig) and class BitsAndBytesLinearMethod (LinearMethodBase).
> * Examples/qlora_inference.py: Demonstration of the use of bitsandbytes, both with and without an adapter.

VirgilG72 commented:
> Hello, I am appreciated for your excellent work. However, I have noticed about that "Also, no TP or PP with QLoRA is supported. It will be considered as the immediate next effort.". So, May I ask when this job is expected to be supported in the future？
> 
> > QLoRA (https://arxiv.org/abs/2305.14314) cuts memory consumption in LLM weight loading without degrading performance. The weights of the basic model , which are quantized into 4 bit using bitsandbytes quantization, pair with a low-rank but higher-precision Low-Rank weight matrix to generate output.
> > This MR is the first step in supporting QLoRA in vLLM. With the PR, the Qlora author's open model on hugging face, such as, is supported:
> > 
> > * https://huggingface.co/timdettmers/qlora-flan-7b (Its corresponding large model is "huggyllama/llama-7b")
> > 
> > User can run with or without a QLoRA adapter.
> > So far, only llama as a basic model is supported. More to come in the future. As explained below, special consideration is made for extensibility to future changes and other models. Also, no TP or PP with QLoRA is supported. It will be considered as the immediate next effort.
> > **Explanation on Changes**
> > Modified files mainly include
> > 
> > * Modify vllm/config.py, vllm/engine/arg_utils.py: Add new CLI parameters for QLoRA/bitsandbytes. The new parameter is:
> >   
> >   * qlora_adapter_name_or_path : the path to the adpater repo. Could be empty.
> > * Modify vllm/model_loader/loader.py: Define a new loader class, which will quantize the weight using bitsandbytes during loading
> > * Modify vllm/model_executor/layers/linear.py: Add the logic of concatenate tensor in bitsandbytes in the weight_loader () function of QKVParallelLinear class and MergedColumnParallelLinear class.
> > 
> > The newly added files are:
> > 
> > * VLLM/model_executor/layers/quantization/bitsandbytes.py: Here, similar to other quantization methods, we define two classes, class BitsAndBytesConfig (QuantizationConfig) and class BitsAndBytesLinearMethod (LinearMethodBase).
> > * Examples/qlora_inference.py: Demonstration of the use of bitsandbytes, both with and without an adapter.

@chenqianfzh

chenqianfzh commented:
> Hello, I am appreciated for your excellent work. However, I have noticed about that "Also, no TP or PP with QLoRA is supported. It will be considered as the immediate next effort.". So, May I ask when this job is expected to be supported in the future？

Actually, TP on bnb has partially done in https://github.com/vllm-project/vllm/pull/5813. Yet the community decided to provide support to FP4 and 8bit first. I will rebase the work PR 5813 after the PR https://github.com/vllm-project/vllm/pull/7445 is merged. 

Hopefully it will be quick. :-)

And, thanks for your checking. I hope I can share my update with you soon.

VirgilG72 commented:
> > Hello, I am appreciated for your excellent work. However, I have noticed about that "Also, no TP or PP with QLoRA is supported. It will be considered as the immediate next effort.". So, May I ask when this job is expected to be supported in the future？
> 
> Actually, TP on bnb has partially done in #5813. Yet the community decided to provide support to FP4 and 8bit first. I will rebase the work PR 5813 after the PR #7445 is merged.
> 
> Hopefully it will be quick. :-)
> 
> And, thanks for your checking. I hope I can share my update with you soon.

Thank you for your reply. Looking forward to your future work!

jeejeelee commented:
> Can a 4-bit LoRA fine-tuned model based on deepseek-ai/DeepSeek-R1-Distill-Qwen-32B be deployed using vLLM without merging the LoRA weights?

Of course

deepanshu12tt commented:
but it is written that it supports only llama based model that's why i am asking?
